src/H5Tconv.c

```diff
@@ -3801,6 +3818,57 @@ H5T__conv_vlen(const H5T_t *src, const H5T_t *dst, H5T_cdata_t *cdata, const H5T
     } /* end switch */
 
 done:
+    /* Release converted elements on error */
+    if (ret_value < 0 && conversions_made) {
+        size_t dest_count;
+
+        /* Set up for first pass to destroy references */
+        if (nelmts < orig_nelmts || (convert_forward && elmtno < safe)) {
+            dest_count = orig_nelmts - nelmts;
+
+            /* Set pointer to correct location, based on direction chosen */
+            if (convert_forward) {
+                d = (uint8_t *)buf;
+                dest_count += elmtno; /* Include partial iteration in first pass, for forward conversions */
+            }
+            else
+                d = (uint8_t *)buf + (nelmts * orig_d_stride);
+
+            /* Destroy references that have already been converted */
+            while (dest_count > 0) {
+                H5_GCC_DIAG_OFF("discarded-qualifiers")
+                H5_CLANG_DIAG_OFF("incompatible-pointer-types-discards-qualifiers")
+                H5T_vlen_reclaim_elmt(d, dst); /* Ignore errors at this point */
+                H5_GCC_DIAG_ON("discarded-qualifiers")
+                H5_CLANG_DIAG_ON("incompatible-pointer-types-discards-qualifiers")
+                d += orig_d_stride;
+                dest_count--;
+            }
+        }
+
+        /* Do any remaining partial iteration, if converting backwards */
+        if (!convert_forward && elmtno < safe) {
+            dest_count = elmtno;
+
+            /* Set pointer to correct location */
+            if (d_stride > 0)
+                d = (uint8_t *)buf + ((nelmts - safe) * orig_d_stride);
+            else
+                d = (uint8_t *)buf + ((nelmts - elmtno) * orig_d_stride);
+
+            /* Destroy references that have already been converted */
+            while (dest_count > 0) {
+                H5_GCC_DIAG_OFF("discarded-qualifiers")
+                H5_CLANG_DIAG_OFF("incompatible-pointer-types-discards-qualifiers")
+                H5T_vlen_reclaim_elmt(d, dst); /* Ignore errors at this point */
+                H5_GCC_DIAG_ON("discarded-qualifiers")
```

Comment:
Same comment here about `H5T_vlen_reclaim_elmt`, as well as the two in `H5T__conv_ref` below (though I'm thinking we need a helper function for this cleanup code.